pex/resolve/locked_resolve.py

```diff
@@ -178,6 +194,138 @@ def select_artifact(self, target):
+
+        if prefer_older_binary:
+            if self.ranked_artifact.rank == other.ranked_artifact.rank:
+                return self if self.version > other.version else other
+            return Rank.select_highest_rank(self, other, lambda ra: ra.ranked_artifact.rank)
+
+        if self.version == other.version:
+            return Rank.select_highest_rank(self, other, lambda ra: ra.ranked_artifact.rank)
+        return self if self.version > other.version else other
+
+
+@attr.s(frozen=True)
+class DownloadableArtifact(object):
+    @classmethod
+    def create(
+        cls,
+        pin,  # type: Pin
+        artifact,  # type: Artifact
+        satisfied_direct_requirements=(),  # type: Iterable[Requirement]
+    ):
+        # type: (...) -> DownloadableArtifact
+        return cls(
+            pin=pin,
+            artifact=artifact,
+            satisfied_direct_requirements=SortedTuple(satisfied_direct_requirements, key=str),
+        )
+
+    pin = attr.ib()  # type: Pin
+    artifact = attr.ib()  # type: Artifact
+    satisfied_direct_requirements = attr.ib(default=SortedTuple())  # type: SortedTuple[Requirement]
+
+
+@attr.s(frozen=True)
+class Resolved(object):
+    @classmethod
+    def create(
+        cls,
+        target,  # type: Target
+        direct_requirements,  # type: Iterable[Requirement]
+        downloadable_requirements,  # type: Iterable[_ResolvedArtifact]
+    ):
+        # type: (...) -> Resolved
+
+        direct_requirements_by_project_name = defaultdict(
+            list
+        )  # type: DefaultDict[ProjectName, List[Requirement]]
+        for requirement in direct_requirements:
+            direct_requirements_by_project_name[ProjectName(requirement.project_name)].append(
+                requirement
+            )
+
+        downloadable_artifacts = []
+        requirement_ranks = []
+        for downloadable_requirement in downloadable_requirements:
+            pin = downloadable_requirement.locked_requirement.pin
+            downloadable_artifacts.append(
+                DownloadableArtifact.create(
+                    pin=pin,
+                    artifact=downloadable_requirement.artifact,
+                    satisfied_direct_requirements=direct_requirements_by_project_name[
+                        pin.project_name
+                    ],
+                )
+            )
+            requirement_ranks.append(downloadable_requirement.ranked_artifact.rank.value)
+
+        average_requirement_rank = sum(requirement_ranks) / float(len(requirement_ranks))
+
+        # N.B.: Lowest rank means highest rank value. I.E.: The 1st tag is the most specific and
+        # the 765th tag is the least specific.
+        largest_value = target.supported_tags.lowest_rank.value
+
+        return cls(
+            target_specificity=(largest_value - average_requirement_rank) / largest_value,
+            downloadable_artifacts=tuple(downloadable_artifacts),
+        )
```

Comment:
This method also seems like it could benefit from a unit test, that we get the math right.